src/contracts/libraries/GPv2Transfer.sol

```diff
@@ -0,0 +1,83 @@
+import "./GPv2SafeERC20.sol";
+
+/// @title Gnosis Protocol v2 Transfers
+/// @author Gnosis Developers
+library GPv2Transfer {
+    using GPv2SafeERC20 for IERC20;
+
+    /// @dev Transfer data.
+    struct Data {
+        address account;
+        IERC20 token;
+        uint256 amount;
+        bool useInternalBalance;
+    }
+
+    /// @dev Ether marker address used to indicate an Ether transfer.
+    address internal constant BUY_ETH_ADDRESS =
+        0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    /// @dev Execute the specified transfers from the specified accounts to a
+    /// single recipient.
+    ///
+    /// This method is used for accumulating user balances into the settlement
+    /// contract.
+    ///
+    /// @param vault The Balancer vault to use.
+    /// @param recipient The single recipient for all the transfers.
+    /// @param transfers The batched transfers to perform.
+    function transferToRecipient(
+        IVault vault,
+        address recipient,
+        Data[] calldata transfers
+    ) internal {
+        // NOTE: Pre-allocate an array of vault balance tranfers large enough to
+        // hold all transfers. This allows us to efficiently batch internal
+        // balance transfers into a single Vault call.
+        IVault.BalanceTransfer[] memory vaultTransfers =
+            new IVault.BalanceTransfer[](transfers.length);
+        uint256 vaultTransferCount = 0;
+
+        for (uint256 i = 0; i < transfers.length; i++) {
+            Data calldata transfer = transfers[i];
+            require(
+                address(transfer.token) != BUY_ETH_ADDRESS,
+                "GPv2: cannot transfer native ETH"
+            );
+
+            if (transfer.useInternalBalance) {
+                IVault.BalanceTransfer memory vaultTransfer =
+                    vaultTransfers[vaultTransferCount++];
+                vaultTransfer.token = transfer.token;
+                vaultTransfer.amount = transfer.amount;
+                vaultTransfer.sender = transfer.account;
+                vaultTransfer.recipient = recipient;
+            } else {
+                transfer.token.safeTransferFrom(
+                    transfer.account,
+                    recipient,
+                    transfer.amount
+                );
+            }
+        }
+
+        if (vaultTransferCount > 0) {
+            // NOTE: Truncate the vault transfers array so that its length is
+            // the number of vault transfers we want to perform. This is done
+            // by setting the array's length in which occupies the first word
+            // in memory pointed to by the `vaultTransfers` memory variable.
+            // <https://docs.soliditylang.org/en/v0.7.6/internals/layout_in_memory.html>
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                mstore(vaultTransfers, vaultTransferCount)
+            }
+            vault.withdrawFromInternalBalance(vaultTransfers);
```

Comment:
This is a bit scary. What happens when the compiler wants to free the memory used by `vaultTransfers`? This is not captured by the test because execution stops after this line is reached. Maybe we could add the usual test to check that memory is freed in full to confirm our expectations.

Otherwise: would it be reasonable instead to have two arrays, `DataInternal` and `DataExternal`, and get the length for free?

Reply:
> Maybe we could add the usual test to check that memory is freed in full to confirm our expectations.

I don't expect memory to be freed. I just expect the array to be shortened to match the number of `vaultTrasnfers` that were actually written. In that regard, I expect the free memory pointer to stay unchanged, and we will have allocated items in memory that we can't access anymore.

> and get the length for free?

At some point, we don't know how many elements in a settlement are internal transfers and how many are external transfers, so we will have to do something like this somewhere.

Reply:
Makes sense! I forgot that memory is never freed.

Reply:
So the idea is that, somewhere in memory (such that `vaultTransfers = x`, and `transfers.length = n`) we allocate something like, where `-` represents an uninitialized `BalanceTransfer`:
```
  x  | x+1 | x+2 | x+3 |       | x+n+1
-----+-----+-----+-----+- ... -+-------
  n  |  -  |  -  |  -  |       |   -
```

Then, when going through the list of transfers, lets say the first (`t0`) and the third (`t2`) use internal balances. Then our memory will look like:
```
  x  | x+1 | x+2 | x+3 |       | x+n+1
-----+-----+-----+-----+- ... -+-------
  n  | t0  | t2  |  -  |       |   -
```

Now, we just want to re-write the length of the array, so it represents the number of elements that were actually written to our "buffer". This effectively makes `x+3 ... x+n+1` inaccessible (represented as `x`) since we no longer have pointers to those memory locations (note, however, that as far as Solidity is concerned, they are still allocated):

```
  x  | x+1 | x+2 | x+3 |       | x+n+1
-----+-----+-----+-----+- ... -+-------
  2  | t0  | t2  |  x  |       |   x
```

In that regard, code that runs after this should "see" that the full `n` elements were allocated, and not the "shortened" allocation.